pandas/core/arrays/categorical.py

```diff
@@ -379,6 +382,110 @@ def __init__(
         self._dtype = self._dtype.update_dtype(dtype)
         self._codes = coerce_indexer_dtype(codes, dtype.categories)
 
+    @classmethod
+    def from_dummies(cls, dummies: "DataFrame", ordered=None):
+        """
+        Create a `Categorical` using a ``DataFrame`` encoding those categories
+        as dummy/ one-hot encoded variables.
+
+        The ``DataFrame`` must be coercible to boolean,
+        and have no more than one truthy value per row.
+        The columns of the ``DataFrame`` become the categories of the `Categorical`.
+        A column whose header is NA will be dropped.
+
+        Parameters
+        ----------
+            dummies : DataFrame of bool-like
+            ordered : bool
+                Whether or not this Categorical is ordered.
+
+        Raises
+        ------
+            ValueError
+                If a sample belongs to >1 category
+
+        Returns
+        -------
+        Categorical
+
+        Examples
+        --------
+        >>> df = pd.DataFrame(
+        ... [[1, 0, 0], [0, 1, 0], [0, 0, 1]], columns=["a", "b", "c"]
+        ... )
+        >>> Categorical.from_dummies(df)
+        [a, b, c]
+        Categories (3, object): [a, b, c]
+        """
+        # GH 8745
+        from pandas import Series
+
+        df = dummies.drop(columns=np.nan, errors="ignore").astype(bool)
+
+        if (df.sum(axis=1) > 1).any():
+            raise ValueError("Some rows belong to >1 category")
+
+        index_into = Series([np.nan] + list(df.columns))
+        mult_by = np.arange(1, len(index_into))
+
+        codes = (df.astype(int) * mult_by).sum(axis=1) - 1
+        codes[codes.isna()] = -1
```

Comment:
I may have misunderstood what's happening here, but would it work to use `pd.factorize`?

Reply:
I may be missing what `factorize` does, but I don't think it would save any effort here -  factorize wants a 1D array (which this isn't until `codes` is instantiated), and then doesn't produce a `Categorical` unless given one, so the last line wouldn't change either.

Reply:
OK yes, I was thinking something like
```python
codes, _ = factorize(df.idxmax(axis=1))
```
, but then it'd be necessary to deal with the case when there's a row without any ones, which'd require more code and would be no shorter/clearer than what you've already got. So, ignore the `factorize` suggestion :)

Could of thiings then:
- is `astype(int)` necessary? It should work without it
- perhaps there would be a comment explaining why you do `-1` at the end of this line
```python
        codes = (df.astype(int) * mult_by).sum(axis=1) - 1
```
?
Now that I've run the code, it's obvious what it does, but it wasn't when I was just reading it

Reply:
You're right, the `astype` wasn't necessary. There's a comment now, best way I could think of explaining it but I can go for prose if you prefer.